### PR TITLE
Towards GPU Compatible BVP Solvers

### PR DESCRIPTION
**Pull request by avik-pal**
* Rolls back the matrix / Array{3} internal representations
* Returns to a vector of arrays representation but performs all operations in-place if possible  
    * Previous implementation was vector of arrays but still reallocated everything
* GPUs should work with the current version (still to be tested)
* Needs a merge of #98 first
* Overall allocations are significantly down (tested against master)

## TODOs

- [ ] Does `f` on GPU work?
    * Really bad performance currently -- BandedMatrix doesn't quite work
    * coloring fails for CuSparse
- [x] Can `u` be anything other than a vector?

### CODE REVIEW

File: .JuliaFormatter.toml [hunk 1]
```diff
@@ -1 +1,2 @@
-style = "sciml"
\ No newline at end of file
+style = "sciml"
+annotate_untyped_fields_with_any = false
```
Comment (ChrisRackauckas) [inline .JuliaFormatter.toml:2]: isn't this already in the sciml format?

Comment (ChrisRackauckas) [reply .JuliaFormatter.toml:2]: It doesn't do that 😅

Comment (avik-pal) [reply .JuliaFormatter.toml:2]: https://github.com/domluna/JuliaFormatter.jl/blob/c72d049210a06e2a404df51998051f3dbb0a123b/src/styles/sciml/pretty.jl#L25

It seems okay to have it as true by default, but if we use something like ConcreteStructs (which IMO we should be using everywhere), then it becomes a problem.

File: src/adaptivity.jl [hunk 2]
```diff
@@ -115,149 +107,153 @@ function redistribute(mesh, Nsub_star::Int, ŝ::AbstractVector{T}, dt) where {T
...
-The input mesh has length of n+1
+    half_mesh!(mesh, mesh_dt)
+    half_mesh!(cache::MIRKCache)
 
-Divide the original subinterval into two equal length subinterval.
+The input mesh has length of `n + 1`. Divide the original subinterval into two equal length
+subinterval. The `mesh` and `mesh_dt` are modified in place.
 """
-@views function half_mesh(mesh::AbstractVector{T}) where {T}
+function half_mesh!(mesh::Vector{T}, mesh_dt::Vector{T}) where {T}
```
Comment (ChrisRackauckas) [inline src/adaptivity.jl:123]: only Vector?

Comment (avik-pal) [reply src/adaptivity.jl:123]: Mesh is controlled internally by us. Irrespective of what the input types are we always construct a Vector. We can relax it, but I am not sure how many types declare stuff like `resize!`

Comment (ChrisRackauckas) [reply src/adaptivity.jl:123]: Won't the mesh need to be a GPU vector for the GPU stuff to work?

Comment (avik-pal) [reply src/adaptivity.jl:123]: Not currently. But again, the code right now is very slow on GPU.

The main issue is we iterate over the mesh to redistribute the new points, so we need to assume fast scalar indexing. We will have to figure out a way to batch the entire computation (which technically is trivial to write a kernel for).

**Pull request merged**